madsim-aws-sdk-s3/src/server/service.rs

```diff
@@ -356,141 +347,133 @@ impl ServiceInner {
-                let end_pos = range_set
-                    .split_once('-')
-                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
-                    .0
-                    .parse::<usize>()
-                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
-
-                object.body.before(end_pos).await
+            let range_set = split.next().ok_or_else(invalid_range)?;
+
+            let (begin_str, end_str) = range_set.split_once('-').ok_or_else(invalid_range)?;
+            let begin_pos = if begin_str.is_empty() {
+                None
             } else {
-                let begin_pos = range_set
-                    .split('-')
-                    .next()
-                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
-                    .parse::<usize>()
-                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
-
-                let end_pos = range_set
-                    .split('-')
-                    .next()
-                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
-                    .parse::<usize>()
-                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
-
-                object.body.between(begin_pos, end_pos).await
+                Some(begin_str.parse::<usize>().map_err(|_| invalid_range())?)
+            };
+            let end_pos = if end_str.is_empty() {
+                None
+            } else {
+                Some(end_str.parse::<usize>().map_err(|_| invalid_range())?)
+            };
+            let body = match (begin_pos, end_pos) {
+                (Some(begin), Some(end)) => object.body.slice(begin..=end),
+                (Some(begin), None) => object.body.slice(begin..),
+                (None, Some(len)) => object.body.slice(object.body.len() - len..),
+                (None, None) => object.body.slice(..),
             };
 
-            Ok(GetObjectOutput { body })
+            Ok(GetObjectOutput { body: body.into() })
         } else if let Some(part_number) = part_number {
-            if part_number >= 0 {
-                let part_number = part_number as usize;
-
-                if part_number < object.body.len().await {
-                    return Ok(GetObjectOutput {
-                        body: object.body.between(part_number, part_number + 1).await,
-                    });
-                }
-            }
-            Err(Error::InvalidPartNumberSpecifier(part_number))
+            if part_number < 0 || part_number as usize >= object.body.len() {
+                return Err(GetObjectError::unhandled(format!(
+                    "invalid part number: {part_number}"
+                )));
+            };
+            let part_number = part_number as usize;
+            Ok(GetObjectOutput {
+                // XXX(wrj): not right?
+                body: object.body.slice(part_number..part_number + 1).into(),
```

Comment:
yes, this is not right. see https://github.com/aws/aws-sdk-cpp/issues/662#issuecomment-1079464053.Maybe we need further test and reserve the multipart information, or just make the action todo

Reply:
I replaced it with `todo!()` to prevent wrong result.